### PR TITLE
[SPARK-21760] [Streaming] Fix for Structured streaming terminates with Exception

### PR DESCRIPTION
**Pull request by rekhajoshm**
## What changes were proposed in this pull request?
Updated serialize call for flush and FSDataOutputStream hflush, and IOUtils as per quick analysis of code

## How was this patch tested?
existing tests

### CODE REVIEW

**Comment by rekhajoshm** [comment]
Seems to be apache spark git/jenkins issue.Please retest after a while.thanks

**Comment by srowen** [comment]
@steveloughran what do you think of this? 

flushing sounds safe but is there a performance impact here if done on every `serialize`?

**Comment by steveloughran** [comment]
Problem here is that a stream which doesn't implement hflush/hsync is required to throw an exception; it's a way of guaranteeing that if hsync/hflush does complete, the action has done what you want - HBase &c utterly depend on this.

The fact that FSDataOutputStream implements Syncable and yet streams it may relay to may not is the whole reason for [HDFS-11644](https://issues.apache.org/jira/browse/HDFS-11644) and the `StreamCapabilities` method. As with Erasure Coding, even HDFS streams may not support hflush/hsync

This patch is at risk of raising an exception whenever it tries to call `hflush()` on non HDFS store or HDFS with Erasure Coding enabled. IF you were targeting Hadoop 2.9+ you could just check `hasCapability("hsync")` use it if present. For Hadoop 2.6+ you'll have to call `out.hflush()` on the first attempt, if any exception (IOE, UnsupportedOperationException, RTE) is raised, catch, swallow and never try to hflush again. 

Sorry, it's messy: its why I'd like that `hasCapability(`) probe up for all features which are only intermittently available. Can complicate caller code if you want to know these things, but stops you getting caught out when you really want to know the durability semantics of the FS.

see also WiP [OutputStream](https://github.com/steveloughran/hadoop/blob/s3/HADOOP-13327-outputstream-trunk/hadoop-common-project/hadoop-common/src/site/markdown/filesystem/outputstream.md)

(thanks for mentioning me BTW; this is one of those things that would probably work well in local tests but blow up in production somewhere)

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/HDFSMetadataLog.scala [hunk 1]
```diff
@@ -123,6 +123,7 @@ class HDFSMetadataLog[T <: AnyRef : ClassTag](sparkSession: SparkSession, path:
           serialize(metadata, output)
           return Some(tempPath)
         } finally {
+          output.hflush()
```
Comment (steveloughran) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/HDFSMetadataLog.scala]: not needed, close will do it.

File: sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/CompactibleFileStreamLog.scala [hunk 2]
```diff
@@ -139,6 +139,9 @@ abstract class CompactibleFileStreamLog[T <: AnyRef : ClassTag](
       out.write('\n')
       out.write(Serialization.write(data).getBytes(UTF_8))
     }
+    if (out.isInstanceOf[FSDataOutputStream]) {
+      out.asInstanceOf[FSDataOutputStream].hflush
+    }
```
Comment (steveloughran) [inline sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/CompactibleFileStreamLog.scala]: Something like 

When opening stream
```scala
var syncable: Option[Syncable]] = if (out.isInstanceOf[FSDataOutputStream]) Some(out.asInstanceOf[FSDataOutputStream]) else None
```

then when at flush time, any flush failure just downgrades to a "don't flush any more"

```scala
try {
syncable.map(_.hflush)
} catch {
case _: Exception =>  syncable = None
}
```

I know, risky,  but I'm not 100% sure that its always UnsupportedOperationException which gets raised. If IO is failing, it'll surface later on anyway.

**Comment by steveloughran** [comment]
BTW, perf wise: hflush() is required to block until the flush has got to the store (visible to others), and with hsync actually  saved to the durable store. So it will take time, but if you want durability, that's that price. Without that hsync call though, there's no guarantee anything will be written to the store. If you need this log to recover from failures: hsync is what you are going to need, even though it currently only works on HDFS and WASB.

**Comment by rekhajoshm** [comment]
Thanks for the good inputs.Closing this PR.

**Comment by steveloughran** [comment]
I think the sync is important, but that you just need to handle the case of "fs doesn't support it".

Thinking about this a bit more, I didn't like my proposed patch. Better to have

* probe for feature after open(), through a check for implementing Syncable, and then calling `hflush()`. It's the lower cost call and if you implement one, you have to implement the other.
* if hflush fails, don't use sync, so set `syncable: Optional<Syncable>` to None
* when checkpointing, go `syncable.map(_.hsync())`. Which is the core of your current patch

you will take a perf hit on the sync, as on HDFS you won't get it returning until it has been written down the entire replication chain. But after that, you've got a guarantee of durability, which is what checkpoints tend to expect...

(side topic: some JIRAs on Flink checkpointing to other stores, especially [FLINK-9061](https://issues.apache.org/jira/browse/FLINK-9061)

**Pull request closed without merging**